cmake/android: build SDL3.jar + SDL3-javadoc + build apk's from tests

madebr opened the pull request:
- Build SDL3.jar with CMake, containing the Android java sources
- Build SDL3-javadoc.jar with CMake, containing the javadoc of the Android java sources
- Build apk's for the SDL tests with CMake. These can also be installed, started and uninstalled using the make tool. Gradle and ndk-build are not involved.

This is also tested on ci


## Description

- These scripts are not written for SDL users, but for the SDL developers.
I believe not needing android studio/gradle to quickly verify things makes developing for Android a more pleasant experience.
No promises are made about API stability of the new cmake functions.
If the name of the CMake project including them does not start with SDL, a warning is displayed.

- The apk's are not built by default. You need to explicitly build a target.

- The CMake script will auto-detect installed Android platforms from the `ANDROID_HOME` environment variable.

- The generated apk's contain native libraries for a single architecture.
So if you want to install them on your current phone, you need to configure with `-DANDROID_ABI=arm64-v8a`.

Example use:
```
# Configure CMake for arm64-v8a
cmake -S . -B cmake-build-android -GNinja -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK_HOME/build/cmake/android.toolchain.cmake  -DANDROID_ABI=arm64-v8a -DSDL_TESTS=ON

# Build native + java libraries (libSDL3.so, SDL3.jar, libtestloopwave.so, libtestsprite.so, ...)
ninja -C cmake-build-android

# Build the apk for testspriteminimal
ninja -C  cmake-build-android testspriteminimal-apk

# Install the apk on your Android mobile
ninja -C  cmake-build-android install-testspriteminimal

# Start testspriteminimal on your Android mobile
ninja -C  cmake-build-android start-testspriteminimal

# Build all apk's
ninja -C cmake-build-android sdl-test-apks

# Install all apk's
ninja -C cmake-build-android install-sdl-test-apks

# Uninstall all apk's
ninja -C cmake-build-android uninstall-sdl-test-apks
```

<details>

<summary>For those interested, the build process implemented in CMake</summary>

<img src="https://user-images.githubusercontent.com/4138939/229972034-a19ce319-945b-4bf1-9a27-4fff83f159b5.png" alt="Process for creating SDL"/>

</details>



## Existing Issue(s)
<!--- If it fixes an open issue, please link to the issue here. -->

ericoporto commented:
Will It still be possible to use Android Studio for development? I really don't want to use binaries from SDL and prefer to keep building from source as I always did. I have an App where I have both the SDL2 stuff for rendering the game "screen" but I use Kotlin for the Android native UI and also for connecting with Android specific services, which would be either cumbersome or impossible using only C/C++ or whatever in the "native" side.

madebr commented:
> Will It still be possible to use Android Studio for development?

Yes. Nothing changes when you call cmake from your gradle script in Android Studio.
The `SDL3::SDL3*` targets remain untouched.

This pr only allows devs to create apk's with only using cmake + extra targets.
It allows them to build the tests with the same *ease* as building executables for psp and n3d, who also need custom build steps.

madebr commented:
> I've tried it and it's works and very convenient and fast to build/run one of the SDL test programs.

Always great to hear it also works on someone else's computer ;)

> maybe:
> 
>     * we don't need the extra activity. just SDLActivity to be as close as possible as user program.

The extra activity is not used when starting the application through a tap. I found it useful to switch from the opengles renderer to the software renderer. It's not ideal though, because the test logs to logcat, and not to the screen. So you need to keep your device connected to your computer to see what's wrong.

>     * copy paste all the issue/doc, in doc/readme-android ?

Sure, as long as it mentions that it does not support external projects.

>     * (eventually a basic shell script that groups all the step to prepare / build / install / run a  test-program (in argument) )

Not 100% sure, because the number of commands is currently already very low:
```
cmake -S SDL -B build
cmake --build build --target sdl-test-apks
cmake --build build --target install-sdl-test-apks
```

1bsyl commented:
Btw, the convenient script, I used. for quickly iterating source/build/install :

```
#!/bin/sh
APP=$1

# Check the error-code is 0, or exit with error message
check_cmd()
{
   RET=$1
   STR=$2
   if [ $1 != 0 ]
   then
      echo "   --> Error: $2"
      exit -1
   fi
}

ninja -C  cmake-build-android ${APP}-apk
check_cmd $? "Build"

ninja -C  cmake-build-android install-${APP}
check_cmd $? "Install"

ninja -C  cmake-build-android start-${APP}
check_cmd $? "Start"
```

madebr commented:
I've updated the documentation.

I've added a convenience target that builds, installs and starts a test. I think that makes a shell script not needed anymore.
```
ninja build-install-start-testsprite
```

Another little change is that `install-testsprite` will (re)build `testsprite-apk`, if it's outdated.

madebr commented on .github/workflows/android.yml:
```diff
@@ -28,6 +28,11 @@ jobs:
           ./build-scripts/androidbuildlibs.sh
+      - uses: actions/setup-java@v3
+        if: ${{ matrix.platform.name == 'CMake' }}
+        with:
+          distribution: 'temurin'
+          java-version: '8'
```
I noticed [this recent commit](https://github.com/libsdl-org/SDL/commit/b701ac026661d2a5be0079ed023bb75d4f651e26) for Java 1.7 support. 
8 is the oldest Java release that still has support (LTS).

madebr replied:
Looks like this fails to create a source jar archive so I bumped it to 11.
log: https://github.com/libsdl-org/SDL/actions/runs/4853935644/jobs/8650710471

madebr replied:
I fixed the original error, but still have to bump the java version because of the following error when running `d8`
```
cd /home/runner/work/SDL/SDL/build/test && /usr/local/lib/android/sdk/build-tools/34.0.0/d8 /home/runner/work/SDL/SDL/build/test/checkkeysthreads.jar /home/runner/work/SDL/SDL/build/SDL3-3.0.0.jar --lib /usr/local/lib/android/sdk/platforms/android-34/android.jar --output /home/runner/work/SDL/SDL/build/test/CMakeFiles/checkkeysthreads-dex.dir
Error: A JNI error has occurred, please check your installation and try again
Error: Exception in thread "main" java.lang.UnsupportedClassVersionError: com/android/tools/r8/D8 has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:473)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:621)
```

madebr commented on test/android/cmake/AndroidManifest.xml.cmake:
```diff
@@ -0,0 +1,70 @@
+
+    <application
+        android:allowBackup="true"
+        android:icon="@mipmap/sdl-test"
+        android:roundIcon="@mipmap/sdl-test_round"
```
I've created a svg approximation of [SDL's logo](https://github.com/libsdl-org/SDL/blob/main/Xcode/SDL/pkg-support/sdl_logo.pdf):

![sdl_logo](https://user-images.githubusercontent.com/4138939/235517567-773c674a-10af-4222-abd8-25c481f02dd0.svg)


![SDL_beta](https://user-images.githubusercontent.com/4138939/235517580-df9af453-d321-4bc0-b8de-cb27293bc133.svg)

madebr commented:
The apks can be downloaded on the github actions page:
https://github.com/libsdl-org/SDL/actions/runs/5437939574?pr=7581

Looks like all providing al apks is a bit heavy (264MB).
Perhaps a subset is useful for testing purposes?

ssrobins commented:
I tried running the commands under 'Example use', but `ninja -C cmake-build-android` didn't produce the `SDL3.jar` and all commands `ninja -C  cmake-build-android testspriteminimal-apk` and later, give me an `ninja: error: unknown target` error. What am I missing?

Does `cmake --install` install everything needed to make an SDL Android app? That's a key feature needed in order to have this 'just work' in the Conan package recipe:
https://github.com/conan-io/conan-center-index/blob/531628920279069a5a5e84a18a91635bd394baec/recipes/sdl/all/conanfile.py#L338

madebr commented:
> The apks can be downloaded on the github actions page: https://github.com/libsdl-org/SDL/actions/runs/5437939574?pr=7581
> 
> Looks like providing all apks is a bit heavy (246MB). Perhaps a subset is useful for testing purposes?

It now only builds testaudiocapture, testcontroller, testmultiaudio and testsprite as an apk.

There's an alternative.
Right now, the a apk is built for every test, which means `libSDL3.so` is installed once for every test and present N times in the apk archive.
It is also possible to build a "mega apk" that contains all tests: 1 libSDL3.so and N entry points for the tests.
This would allow to provide all tests with a reasonable download size.

madebr commented:
> Does `cmake --install` install everything needed to make an SDL Android app? That's a key feature needed in order to have this 'just work' in the Conan package recipe: https://github.com/conan-io/conan-center-index/blob/531628920279069a5a5e84a18a91635bd394baec/recipes/sdl/all/conanfile.py#L338

Yes.
```
└── share
    ├── java
    │   └── SDL3
    │       ├── SDL3-3.0.0.jar
    │       ├── SDL3-3.0.0-sources.jar
    │       └── SDL3.jar -> SDL3-3.0.0.jar
    ├── javadoc
    │   └── SDL3
    │       ├── allclasses-frame.html
    │       ├── ...
...
```

backrock commented:
Hi, where is cmake-build-android? or what is cmake-build-android?

madebr commented:
> Hi, where is cmake-build-android? or what is cmake-build-android?

Not sure what you're asking, but the cmake code used to build android can be found in 'cmake/android' and at the bottom of 'test/CMakeLists.txt'

madebr commented:
Oh, 'cmake-build-android' is the name of the build directory.

backrock commented:
> Oh, 'cmake-build-android' is the name of the build directory.

thanks
